### ISSUE TITLE
TSC failing to deduce this inside a method definition when calling an inherited method with a distributive parameter type

### ISSUE THREAD

**Issue by jgonggrijp** [microsoft/TypeScript#42413]
# Bug Report

### 🔎 Search Terms

cannot deduce this

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about `this`.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?module=1&ts=3.3.3#code/C4TwDgpgBAygrgIwG4EsIHcByBDAthAZwB4AVAGigFUoIAPYCAOwBMCoBrCEAewDMoSUALwcufAQD5hVGvSasocRu0bd0jKAH4BAbUoBdWQxZtGcXAggAnLTIBcURhCTWoDpy6sBuAFA-mEADGADbYVtAh2ARsAGLc3FAA3j5QqVDY7uaW3n5pUMwoBGDcBBAwKLhgwRAAFLQOnDz8wAAWhQCUDkjcKMy+eQVFJRAAwtWedQ7wyGhYeIRErYUSnVDdvb4Avn6R0VAAQmFG8rHxSSlpCACEmRbW-WmBN45Z9xepzM8EwFYojADmvneUF4jBq7XOeTySwIADpBsVSuVKtUagAibBo9oPKFQGHwwqIsoVKq1NEILFA3GpAD0NKgBBa3DgwWYUHQ3Cs7GB0LacIRwzGzms6Mx2J5aXxAtKQom5MpuWpdPy3EI7M53OpUBqMPSbEOVnaBKGMvGIoxCq1Or5eoOYSN0tGZqs6Ip4ry222PmVCDgwDxNo5XLYV00-iCoXCUEC3EY3ygCDCDgNvkTVmNRNl5rFXiAA)

Credit due to @webstrand for reducing this phenomenon to its absolute essence.

### 💻 Code

```ts
type SubviewNames<T, U extends keyof T = keyof T> = U extends unknown ? T[U] extends number ? U : never : never;

declare class Foo {
    a: number;

    disposeSimple(x: keyof this): void;
    disposeClever(x: SubviewNames<this>): void;
}

class Bar extends Foo {
    b!: number;
    c!: number;
    d!: string;

    fn() {
        this.disposeSimple("a");
        this.disposeSimple("b");

        // should work
        this.disposeClever("a");
        this.disposeClever("b");

        // does work
        (this as Bar).disposeClever("a");
        (this as Bar).disposeClever("b");
    }
}

// but this works !?
declare const bar: Bar;
bar.disposeClever("a");
```

### 🙁 Actual behavior

On the lines marked `// should work`, TSC reports the error `Argument of type '"a"' is not assignable to parameter of type 'SubviewNames<this, keyof this>'.` Likewise for `type '"b"'`. This is wrong, because `"a"` and `"b"` evidently extend `keyof Bar` and both `Bar["a"]` and `Bar["b"]` are `number`. In addition, it is inconsistent with the lines above and below.

### 🙂 Expected behavior

That all lines in the example compile without errors, because all types check.

**Comment by RyanCavanaugh** [comment]
`SubviewNames<this>` can't be resolved into a concrete type because a subclass of `Bar` could produce any kind of arbitrary value from this. From a higher-order reasoning perspective we (as humans) can tell that `"a"` and `"b"` *should* be assignable to this type, but this requires knowing that:
 * Subclasses of `Bar` will be assignable to it
 * Nothing in the definition of `SubviewNames` produces contravariant behavior on `keyof T`
 * Nothing in the definition of `SubviewNames` produces contravariant behavior on `T[U]`

Basically we lack a mechanism to evaluate an assignment in this direction and there isn't really a plausible way to add that without new kinds of higher-order reasoning about these constructs

**Comment by jgonggrijp** [comment]
@RyanCavanaugh Thanks for clarifying. Could you expand a bit on these new kinds of higher-order reasoning? In particular, do you think these should be added in the future, or if you're still undecided about that, what does it depend on?
